File: src/ast/tree/statement/return.ts
import type { Pass } from "../../../pass/pass";
import type { Block } from "../../node/block";
import type { Dependent, Expression } from "../../node/expression";
import { ExecutionResult, Statement } from "../../node/statement";

export class Return extends Statement {
    constructor(public value: Dependent<any>, parent: Block | null) {
        super(parent);
    }

    public execute(): ExecutionResult {
        this.parent!.returnValue = this.value.get();
        return ExecutionResult.Return;
    }

    public visit(pass: Pass): void {
        pass.visitReturn(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/statement/Return.kt
package me.honkling.miniscript.parser.ast.statement

import me.honkling.miniscript.parser.ast.Block
import me.honkling.miniscript.parser.ast.expression.Expression
import me.honkling.miniscript.pass.Pass

class Return(
    val expression: Expression<*>,
    parent: Block
) : Statement(parent) {
    override fun execute(): ExecutionResult {
        parent.returnValue = expression.get()
        return ExecutionResult.Return
    }

    override fun accept(pass: Pass) {
        pass.visitReturn(this)
    }
}